### PR TITLE
chore: Make use of pattern matching on singleton type instead.

### PR DESCRIPTION
**Pull request by He-Pin**
More info: @som-snytt point out in https://scala-lang.org/files/archive/spec/2.13/08-pattern-matching.html#type-patterns see singleton.

Thanks @som-snytt for this.

continue : https://github.com/apache/incubator-pekko/pull/983

### CODE REVIEW

**Comment by mdedetrich** [comment]
Build is failing

**Comment by He-Pin** [comment]
@mdedetrich should we remove `infer-any`

**Comment by som-snytt** [comment]
I just tried it. I forgot about https://github.com/scala/bug/issues/12569

which I think it's just necessary to `@nowarn`.  Alternatively, `-Wconf` something something.

That is some kind of pattern matching bug, but it's "just a warning".

**Comment by som-snytt** [comment]
Aliasing it and hiding the type seems to work without nowarn.
```
val fallback: AnyRef = NotApplied.asInstanceOf[AnyRef]
          pf.applyOrElse(grab(in), NotApplied) match {
            case _: fallback.type => completeStage()
```

**Comment by mdedetrich** [comment]
Hmm, with all of the workarounds not sure if this is that much better than whats currently there

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/Ops.scala
Comment (He-Pin) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/Ops.scala:357]: It's better which cover the error case. @mdedetrich

**Comment by He-Pin** [comment]
> val fallback: AnyRef = NotApplied.asInstanceOf[AnyRef]
>           pf.applyOrElse(grab(in), NotApplied) match {
>             case _: fallback.type => completeStage()

@som-snytt what about 
```
          pf.applyOrElse(grab(in), NotApplied: AnyRef) match {
            case _: NotApplied.type => completeStage()
```

**Comment by He-Pin** [comment]
I just run the bench
image

@mdedetrich @som-snytt cc, performance is the same as `eq`.

BTW, the sbt-jmh issue on Windows been verified fixed, thanks.

**Comment by He-Pin** [comment]
```
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                          Mode  Cnt         Score        Error  Units
[info] CollectBenchmark.benchNewCollect  thrpt   10  15246702.956 ± 127702.345  ops/s
[info] CollectBenchmark.benchOldCollect  thrpt   10  14828734.816 ± 112590.257  ops/s
[success] Total time: 773 s (12:53), completed 2024年1月23日 下午10:41:06
```

**Comment by He-Pin** [comment]
@mdedetrich @pjfanning ping, what's your option on this, I tested seems the performance is the same as my previous commit.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala
Comment (Roiocam) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:54]: Will these comments be outdated after this PR?

**Review by Roiocam — COMMENTED** [review]
```
[info] Benchmark                            Mode  Cnt        Score        Error  Units
[info] CollectBenchmark.benchNewCollect    thrpt    3  9147229.404 ± 255711.579  ops/s
[info] CollectBenchmark.benchOldCollect    thrpt    3  8889505.583 ± 985315.585  ops/s
[info] CollectBenchmark.oldIFACMPNE        thrpt    3  8899066.666 ± 271187.167  ops/s
```
Just a quick bench, looks great.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:54]: I think the it's still apply, maybe drop the :
>//   If you interest, you can check the associated PR for this change and the current implementation of `scala.collection.IterableOnceOps.collectFirst`.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala
Comment (jrudolph) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:56]: Why not match on the singleton instance instead? In any case, it seems uncommon to match on the singleton type.

It's not even easy to say if matching on the type or the reference is supposed to be faster in general, and maybe the Scala compiler should make the right choice. (On one hand, reference comparison is as fast as it gets, on the other hand, getting the singleton instance reference is at least a static field lookup which might not be faster than comparing types, would depend on the JIT compiler).

Comment (mdedetrich) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:56]: > Why not match on the singleton instance instead? In any case, it seems uncommon to match on the singleton type.

I am under the impression it only works if you match on the type and not on the instance, @som-snytt can you confirm?

By only works I mean producing the optimal bytecode which was the point of the change in the first place.

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:56]:
```scala
  def collect[A](a: A, pf: PartialFunction[A, A]): A = {
    pf.applyOrElse(a, NotApplied) match {
      case NotApplied => a
      case _          => a
    }
  }

  def collect2[A](a: A, pf: PartialFunction[A, A]): A = {
    pf.applyOrElse(a, NotApplied) match {
      case _: NotApplied.type => a
      case _                  => a
    }
  }
```

ByteCode:
image

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:56]: @mdedetrich @jrudolph I think just as @som-snytt said , the current version should generate less bytecode.

**Comment by He-Pin** [comment]
final results:
Jmh/run -i 10 -wi 10 -f1 -t1 .CollectBenchmark.
```
[info] Benchmark                                      Mode  Cnt                  Score                 Error  Units
[info] CollectBenchmark.benchCollectOnSingleton      thrpt   10  17608004768421276.000 ± 192500018073910.500  ops/s
[info] CollectBenchmark.benchCollectOnSingletonType  thrpt   10  17912827936113308.000 ± 239976332956371.340  ops/s
[info] CollectBenchmark.benchNewCollect              thrpt   10           15283510.041 ±          460993.700  ops/s
[info] CollectBenchmark.benchOldCollect              thrpt   10           14845265.443 ±          178628.150  ops/s
```

**Comment by He-Pin** [comment]
After carefully bench and diff the bytecode, I think this one should be ready to go.

**Review by mdedetrich — APPROVED** [review]
lgtm from my side, would also wait for another review from someone else.

**Review by pjfanning — APPROVED** [review]
lgtm

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala
Comment (som-snytt) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala:56]: The clunky pattern is just to avoid a clunky `eq`.

You might use a local val to avoid "cost of obtaining the value".

This is like pattern match on List with Nil case, where match on `::` first, then just `case nil`. It's nice if you have a type to check.

An example where just "obtaining the value" is expensive is `NoSymbol` in the compiler, because of the "cake". There are edits trying to choose between `eq` and `==` but that is swamped by the fetch. It is outweighed by style concerns.

If you can measure a difference reliably, then of course you are equipped to choose.

**Pull request merged**
